File: java/leet_code/881.java
/**
 * This class provides a solution for the problem of determining the minimum number of rescue boats required.
 * The problem is solved using a bucket sort approach, where the weight of each person is used as an index into a bucket array.
 * The solution iterates over the bucket array from both ends, pairing the lightest and heaviest people together in the same boat where possible.
 */
class Solution {

    /**
     * Calculates the minimum number of rescue boats required to save all people.
     *
     * @param people An array of integers representing the weight of each person. Each weight is a positive integer.
     * @param limit  The maximum weight that a single boat can carry. It is a positive integer.
     *
     * @return The minimum number of boats required to save all people. It is a positive integer.
     */
    public int numRescueBoats(int[] people, int limit) {
        // Create a bucket array to hold the count of people with each weight
        int[] buckets = new int[limit + 1];

        // Populate the bucket array with the count of people with each weight
        for (int weight : people) {
            buckets[weight]++;
        }

        // Initialize pointers to the start and end of the bucket array
        int start = 0;
        int end = buckets.length - 1;

        // Initialize the count of boats required
        int boats = 0;

        // Iterate over the bucket array from both ends
        while (start <= end) {
            // Move the start pointer to the next non-empty bucket
            while (start <= end && buckets[start] <= 0) start++;
            // Move the end pointer to the previous non-empty bucket
            while (start <= end && buckets[end] <= 0) end--;

            // If both pointers have reached empty buckets, break the loop
            if (buckets[start] <= 0 && buckets[end] <= 0) {
                break;
            }

            // Increment the count of boats required
            boats++;

            // If the sum of the weights at the start and end pointers is less than or equal to the limit,
            // decrement the count at the start pointer, as the person at the start can be paired with the person at the end
            if (start + end <= limit) {
                buckets[start]--;
            }

            // Decrement the count at the end pointer, as the person at the end is always placed in a boat
            buckets[end]--;
        }

        // Return the minimum number of boats required
        return boats;
    }
}

File: scala/leet_code/881.scala
/** This module provides a solution to the problem of determining the minimum number of rescue boats required to save
  * people. Each person has a weight, and each boat can carry a maximum weight. The goal is to minimize the number of
  * boats used.
  */

object Solution {

  /** Function to calculate the minimum number of rescue boats required.
    *
    * @param people
    *   an array of integers representing the weights of the people.
    * @param limit
    *   an integer representing the maximum weight a boat can carry.
    * @return
    *   the minimum number of boats required to save all people.
    */
  def numRescueBoats(people: Array[Int], limit: Int): Int = {

    /** Recursive function to calculate the number of boats required.
      *
      * @param max
      *   an integer representing the index of the heaviest person not yet assigned to a boat.
      * @param min
      *   an integer representing the index of the lightest person not yet assigned to a boat.
      * @param boats
      *   an integer representing the number of boats used so far.
      * @return
      *   the number of boats required to save all people.
      */
    @annotation.tailrec
    def plusOneBoat(max: Int, min: Int, boats: Int): Int =
      if (min < max) boats
      else if (people(max) + people(min) <= limit)
        plusOneBoat(max + 1, min - 1, boats + 1)
      else plusOneBoat(max + 1, min, boats + 1)

    people.sortInPlace()(Ordering.Int.reverse)
    plusOneBoat(0, people.length - 1, 0)
  }
}
